### PR TITLE
Feature: Add pprof endpoint

### PR DESCRIPTION
**Pull request by ImpSy**
## Purpose of this PR

This PR aim to allow access to the pprof server expose by the controller-runtime manager

This could help pinpoint issue with profiling while also opening the way for [Profile-guided optimization](https://go.dev/blog/pgo) build of the operator

**Proposed changes:**
- add `--pprof-bind-address` option to the controller
- add the associated values to the helm chart

## Change Category
Indicate the type of change by marking the applicable boxes:

- [ ] Bugfix (non-breaking change which fixes an issue)
- [X] Feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that could affect existing functionality)
- [ ] Documentation update

## Checklist
Before submitting your PR, please review the following:

- [X] I have conducted a self-review of my own code.
- [ ] I have updated documentation accordingly.
- [ ] I have added tests that prove my changes are effective or that my feature works.
- [X] Existing unit tests pass locally with my changes.

### CODE REVIEW

**Comment by ChenYi015** [comment]
@ImpSy Thanks for the contribution to add the pprof endpoint! Could you sign off your commits by `git commit -s -m "commit message"`?

File: ADOPTERS.md
Comment (ChenYi015) [inline ADOPTERS.md:17]: I am wondering why is Data Mechanics deleted? It would be better to raise a dedicated PR to add your company to adopters list.

File: charts/spark-operator-chart/templates/controller/_helpers.tpl
Comment (ChenYi015) [inline charts/spark-operator-chart/templates/controller/_helpers.tpl:107]:
```suggestion
{{/*
Create the name of the service used by controller
*/}}
{{- define "spark-operator.controller.serviceName" -}}
{{ include "spark-operator.controller.name" . }}-svc
{{- end -}}

```

What do you think about renaming the service for controller, so it can be used to expose metric, pprof and other endpoints?

File: charts/spark-operator-chart/templates/controller/deployment.yaml
Comment (ChenYi015) [inline charts/spark-operator-chart/templates/controller/deployment.yaml:103]: Would you like to contribute some helm unit tests to ensure the helm templated can be rendered as expected?

File: ADOPTERS.md
Comment (ImpSy) [reply ADOPTERS.md:17]: Sorry I thought it was clear in the commit msg

Datamechanics has been acquired 3 years ago and the last remaining project has been discontinued in May

The new join product lives under Spot by Netapp

File: charts/spark-operator-chart/templates/controller/_helpers.tpl
Comment (ImpSy) [reply charts/spark-operator-chart/templates/controller/_helpers.tpl:107]: That's actually how we implemented it on our fork

Since lots of things have changed in the repo with the controller-runtime pr I did the minimum effort solution

I'll rename the file to service.yaml and accept your change

File: charts/spark-operator-chart/templates/controller/deployment.yaml
Comment (ImpSy) [reply charts/spark-operator-chart/templates/controller/deployment.yaml:103]: Ok 👍

**Comment by ImpSy** [comment]
@ChenYi015 I've done the changed asked and signed my commits, if you want to take a look

File: ADOPTERS.md
Comment (ChenYi015) [reply ADOPTERS.md:17]: Got it. Would you like to raise a dedicated PR to replace Data Mechanics with Sprt by Netapp?

File: charts/spark-operator-chart/tests/controller/deployment_test.yaml
Comment (ChenYi015) [inline charts/spark-operator-chart/tests/controller/deployment_test.yaml:591]:
```suggestion
          path: spec.template.spec.containers[?(@.name=="spark-operator-controller")].ports
```

File: charts/spark-operator-chart/tests/controller/deployment_test.yaml
Comment (ChenYi015) [inline charts/spark-operator-chart/tests/controller/deployment_test.yaml:582]:
```suggestion
  - it: Should add pprof ports if `controller.pprof.enable` is set to `true`
```

File: charts/spark-operator-chart/tests/controller/service_test.yaml
Comment (ChenYi015) [inline charts/spark-operator-chart/tests/controller/service_test.yaml:34]: The [e2e test CI](https://github.com/kubeflow/spark-operator/actions/runs/10847662027/job/30103203369?pr=2164) has failed for that the controller service cannot be created with empty port lists. I think we should not create service for controller if pprof is disabled.

**Review by ChenYi015 — CHANGES_REQUESTED** [review]

**Comment by ChenYi015** [comment]
> @ChenYi015 I've done the changed asked and signed my commits, if you want to take a look

@ImpSy I have left some comments. The e2e test CI failed due to the controller service has empty ports list. Besides, the DCO check fails, so there must be some commits did not get signed off with `git commit -s -m "message"`.

File: ADOPTERS.md
Comment (ImpSy) [reply ADOPTERS.md]: I've created a PR for this -> https://github.com/kubeflow/spark-operator/pull/2168

**Review by ChenYi015 — APPROVED** [review]

**Comment by ChenYi015** [comment]
Thanks for the contribution! LGTM, will wait for another approval @yuchaoran2011 @vara-bonthu @jacobsalway .

File: cmd/operator/controller/start.go
Comment (jacobsalway) [inline cmd/operator/controller/start.go:166]: Since this flag is passed from `controller.pprof.port`, to me the suggestion in the flag help to use port 6060 seems unnecessary.

**Comment by jacobsalway** [comment]
Thanks for your contribution @ImpSy. I've tested this locally on Kind and it works as intended. I can curl the profiles and analyse them using `go tool pprof`. I'm sure this will prove useful when we analyse operator performance.

I've left one comment on the flag help string but it's more of a nit. The rest of the PR looks good to me.

**Comment by ImpSy** [comment]
@jacobsalway The CLI arg description has been changed

**Review by jacobsalway — APPROVED** [review]

**Comment by jacobsalway** [comment]
@ChenYi015 has already approved the PR and is an approver, so I'm gonna add the label

/lgtm

**Pull request merged**
